pkg/bucket/lifecycle/transition.go

```diff
@@ -18,25 +18,151 @@ package lifecycle
+	if tDate.Time.IsZero() {
+		return nil
+	}
+	return e.EncodeElement(tDate.Format(time.RFC3339), startElement)
+}
+
+// TransitionDays is a type alias to unmarshal Days in Transition
+type TransitionDays int
+
+// UnmarshalXML parses number of days from Transition and validates if
+// >= 0
+func (tDays *TransitionDays) UnmarshalXML(d *xml.Decoder, startElement xml.StartElement) error {
+	var numDays int
+	err := d.DecodeElement(&numDays, &startElement)
+	if err != nil {
+		return err
+	}
+	if numDays < 0 {
+		return errTransitionInvalidDays
+	}
+	*tDays = TransitionDays(numDays)
+	return nil
+}
+
+// MarshalXML encodes number of days to expire if it is non-zero and
+// encodes empty string otherwise
+func (tDays TransitionDays) MarshalXML(e *xml.Encoder, startElement xml.StartElement) error {
+	if tDays == 0 {
+		return nil
+	}
+	return e.EncodeElement(int(tDays), startElement)
+}
+
 // Transition - transition actions for a rule in lifecycle configuration.
 type Transition struct {
-	XMLName      xml.Name `xml:"Transition"`
-	Days         int      `xml:"Days,omitempty"`
-	Date         string   `xml:"Date,omitempty"`
-	StorageClass string   `xml:"StorageClass"`
+	XMLName      xml.Name       `xml:"Transition"`
+	Days         TransitionDays `xml:"Days,omitempty"`
+	Date         TransitionDate `xml:"Date,omitempty"`
+	StorageClass string         `xml:"StorageClass,omitempty"`
+
+	set bool
 }
 
-var errTransitionUnsupported = Errorf("Specifying <Transition></Transition> tag is not supported")
+// MarshalXML encodes transition field into an XML form.
+func (t Transition) MarshalXML(enc *xml.Encoder, start xml.StartElement) error {
+	if !t.set {
+		return nil
+	}
+	type transitionWrapper Transition
+	return enc.EncodeElement(transitionWrapper(t), start)
+}
 
-// UnmarshalXML is extended to indicate lack of support for Transition
-// xml tag in object lifecycle configuration
-func (t Transition) UnmarshalXML(d *xml.Decoder, start xml.StartElement) error {
-	return errTransitionUnsupported
+// UnmarshalXML decodes transition field from the XML form.
+func (t *Transition) UnmarshalXML(d *xml.Decoder, startElement xml.StartElement) error {
+	type transitionWrapper Transition
+	var trw transitionWrapper
+	err := d.DecodeElement(&trw, &startElement)
+	if err != nil {
+		return err
+	}
+	*t = Transition(trw)
+	t.set = true
+	return nil
 }
 
-// MarshalXML is extended to leave out <Transition></Transition> tags
-func (t Transition) MarshalXML(e *xml.Encoder, start xml.StartElement) error {
+// Validate - validates the "Expiration" element
+func (t Transition) Validate() error {
+	if !t.set {
+		return nil
+	}
+
+	if t.IsDaysNull() && t.IsDateNull() {
+		return errXMLNotWellFormed
+	}
+
+	// Both transition days and date are specified
+	if !t.IsDaysNull() && !t.IsDateNull() {
+		return errTransitionInvalid
+	}
+	if t.StorageClass != "" {
+		if _, err := madmin.ParseARN(t.StorageClass); err != nil {
```

Comment:
```suggestion
		if _, err := madmin.ParseARN(t.StorageClass); err != nil {
```
Can ParseARN not return `errXMLNotWellFormed` ?

Reply:
can we not use errMalformedXML a much nicer way to return an error. NOTE `api-errors.go` indeed handle `*xml.UnmarshalError` error returned.

Reply:
Also this looks like very MinIO specific are you sure we should re-purpose StorageClass as ARNs?  why not have them labeled to something more humanly readable? so that this parsing can be avoided?

Reply:
we can add a --label flag in `mc admin bucket remote add` that could be used in `mc ilm add` for storage-class to make it more compatible.